clippy_lints/src/mut_key.rs

```diff
@@ -103,30 +103,63 @@ fn check_sig<'tcx>(cx: &LateContext<'tcx>, item_hir_id: hir::HirId, decl: &hir::
+        if is_keyed_type && is_mutable_type(cx, substs.type_at(0), span) {
             span_lint(cx, MUTABLE_KEY_TYPE, span, "mutable key type");
         }
     }
 }
 
 fn is_mutable_type<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>, span: Span) -> bool {
     match *ty.kind() {
-        RawPtr(TypeAndMut { ty: inner_ty, mutbl }) | Ref(_, inner_ty, mutbl) => {
-            mutbl == hir::Mutability::Mut || is_mutable_type(cx, inner_ty, span)
-        },
+        // Raw pointers are never considered 'mutable' by this lint, since there are two cases:
+        // 1. The raw pointer is hashed/ordered directly (e.g. in `HashSet<*const Cell<usize>>` or
+        //    `HashSet<*mut usize>`.  These impls depend only on the memory **address** of the
+        //    referred value, rather than the value itself.  Therefore, checking the contained type
+        //    causes a false positive.  Links (links refer to Rust version 1.54):
+        //      - clippy issue: https://github.com/rust-lang/rust-clippy/issues/6745
+        //      - hash impl: https://github.com/rust-lang/rust/blob/1.54.0/library/core/src/hash/mod.rs#L717-L736
+        //      - ord impls:
+        //        - https://github.com/rust-lang/rust/blob/master/library/core/src/ptr/const_ptr.rs#L1030-L1043
+        //        - https://github.com/rust-lang/rust/blob/master/library/core/src/ptr/mut_ptr.rs#L1345-L1357
+        //
+        // 2. The raw pointer is contained inside an ADT.  Using `#[derive(Hash)]` will call
+        //    `<*const T as Hash>::hash()`, which will have the same behaviour as case (1).  If the
+        //    ADT uses a custom `impl Hash` which (unsafely) dereferences the raw pointer, then
+        //    they are on their own - checking this in the general case would be impossible.
+        RawPtr(..) => false,
+        Ref(_, inner_ty, mutbl) => mutbl == hir::Mutability::Mut || is_mutable_type(cx, inner_ty, span),
         Slice(inner_ty) => is_mutable_type(cx, inner_ty, span),
         Array(inner_ty, size) => {
             size.try_eval_usize(cx.tcx, cx.param_env).map_or(true, |u| u != 0) && is_mutable_type(cx, inner_ty, span)
         },
         Tuple(..) => ty.tuple_fields().any(|ty| is_mutable_type(cx, ty, span)),
-        Adt(..) => {
-            !ty.has_escaping_bound_vars()
-                && cx.tcx.layout_of(cx.param_env.and(ty)).is_ok()
-                && !ty.is_freeze(cx.tcx.at(span), cx.param_env)
+        Adt(def, substs) => {
+            // `LinkedList`, `Vec`, `VecDeque`, `BTreeSet`, `BTreeMap` all implement `Hash` by
+            // hashing their contents (`BTreeMap` hashes both its keys and values).  All of these
+            // store their contents behind (at least) one level of indirection, meaning that
+            // checking the `Freeze` trait for the type alone will cause false negatives for
+            // obviously bad cases like `HashSet<Vec<Cell<usize>>>`.  So, we handle these
+            // collections explicitly.
```

Comment:
This list should also include Option, Result, HashMap, HashSet. `Option` and `Result` don't have any "indirection" so I'm not sure about that part. I would just say "generic containers where the contained value(s) are hashed and can be modified through interior mutability".

Reply:
We don't need `Hash(Map|Set)`, because they don't implement either of `Hash` or `Ord` and can't be used as keys in the types which trigger this lint.  We do need a recursive case for `Option` and `Result` - cases like `Option<Cell<usize>>` would lint anyway but more complex cases like `Option<Vec<Cell<usize>>>` or `Option<&mut usize>` would not lint unless we recurse.  I've also added tests for these cases.

I've added simpler comment, but one which mentions that we care about both `Hash` (for `Hash(Map|Set)`) _and_ `Ord` (for `BTree(Map|Set)`).

Reply:
Update: I've also added smart pointers like `Box`, `Arc`, `Rc`, `cell::{Ref, RefMut}`, etc.